src/ucp/core/ucp_context.c

```diff
@@ -864,63 +902,109 @@ static int ucp_is_resource_in_device_list(const uct_tl_resource_desc_t *resource
+        return is_allow_mode;
     }
 
-    /* check aliases */
     for (alias = ucp_tl_aliases; alias->alias != NULL; ++alias) {
-        /* If an alias is in the list and the transport belongs this alias,
-         * enable/disable the transport (according to the list mode)
-         */
-        alias_arr_count = ucp_tl_alias_count(alias);
-        snprintf(info, sizeof(info), " for alias '%s'", alias->alias);
-        dummy_mask      = 0;
-        tmp_rsc_flags   = 0;
-        tmp_tl_cfg_mask = 0;
-        dummy_flags     = 0;
-        if (ucp_tls_array_is_present(alias->tls, alias_arr_count, tl_name, info,
-                                     &tmp_rsc_flags, &dummy_mask)) {
-            if (ucp_config_is_tl_name_present(allow_list, alias->alias, 1,
-                                              &dummy_flags, &tmp_tl_cfg_mask)) {
-                if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW) {
-                    *rsc_flags   |= tmp_rsc_flags;
-                    *tl_cfg_mask |= tmp_tl_cfg_mask;
-                    return 1;
-                } else if (!(tmp_rsc_flags & UCP_TL_RSC_FLAG_AUX)) {
-                    /* Disable the transport only if it is not used in the alias
-                     * as an auxiliary transport. */
-                    return 0;
-                }
+        alias_arr_count   = ucp_tl_alias_count(alias);
+        search_result_tmp = ucp_tls_array_search(alias->tls, alias_arr_count,
+                                                 tl_name, &dummy_mask);
+        /* Any combination of 'aux' suffix may enable transport with
+         * appropriate resource flag. The following patterns are valid:
+         * UCX_TLS=...tl_name:aux...; UCX_TLS=...alias:aux,...
+         * But only transport mentioned without 'aux' suffix in the alias are
+         * considered in case of deny list. E.g:
+         * alias = { tl_name_1, tl_name_2:aux }, UCX_TLS=^...alias,... should
+         * not disable tl_name_2. */
+        if (ucp_tls_array_found(search_result_tmp) &&
+            ucp_check_resource_in_transports_list(allow_list, alias->alias, 1,
+                                                  search_stop_mask,
+                                                  &search_result,
+                                                  tl_cfg_mask)) {
+            return is_allow_mode;
+        } else if (ucp_enable_tl_aux(is_allow_mode, search_result_tmp)) {
+            search_result_tmp = ucp_config_tl_name_search(allow_list,
+                                                          alias->alias, 1,
+                                                          tl_cfg_mask);
+            if (ucp_tls_array_found(search_result_tmp)) {
+                search_result |= UCP_TLS_ARRAY_SEARCH_RESULT_FOUND_AUX;
             }
         }
     }
 
-    return allow_list->mode == UCS_CONFIG_ALLOW_LIST_NEGATE;
+    if (ucp_enable_tl_aux(is_allow_mode, search_result) ||
+        ucp_tls_array_found(search_result)) {
```

Comment:
can we really reach this point if `ucp_tls_array_found(search_result)` is true?
I would expect we exited the function in line 956 because search_stop_mask always contains UCP_TLS_ARRAY_SEARCH_RESULT_FOUND

Reply:
Yes, e.g: `UCX_TLS=^ud`. `ucp_tls_array_found(search_result)` is true. But we still need to enable UD as auxiliary transport.